feat: New `WithMaxCost` option for custom cache capacity management strategies

dadrus opened the pull request:
closes #135 

This PR introduces a new `WithMaxCost` option, enabling the restriction of the cache's capacity based on factors beyond just the number of items it stores. This opens up the possibility for implementing custom strategies. For example, one such strategy could involve setting a memory limit for the cache, as demonstrated in the following snippet.

```go
import (
    "github.com/jellydator/ttlcache"
    "github.com/DmitriyVTitov/size"
)

func main() {
    cache := ttlcache.New[string, string](
        ttlcache.WithMaxCost[string, string](5120, func(item *ttlcache.Item[string, string]) uint64 {
            return uint64(size.Of(item))
        }), 
    )

    cache.Set("first", "value1", ttlcache.DefaultTTL)
}
```

With this configuration in place, if an existing item is updated with a new value or a new item is added, causing memory exhaustion (5KiB here), older items will be evicted. In such cases, the eviction reason will be `EvictionReasonMaxCostExceeded`.

dadrus committed:
new option to define the max memory the cache can use in bytes

dadrus committed:
cache.set updated to cope with memory limitations

dadrus commented:
@davseby , @swithek: As you can see in the PR description, tests is an open topic. I actually started implementing those locally, but run in some limitations related to the structure of the available tests (they share too much state in https://github.com/dadrus/ttlcache/blob/6e92deb3b7e755ab332e11670c82b0290cab626f/cache_test.go#L261-L326). Therefore, I would like to discuss a couple of topics with you first:
* Naming of the option and properties I introduced in the PR.
* Making the tests more future proof - I think it would be easier to maintain the tests if each test case could implement its own setup and assertions, similar to the implementation, you can find here: https://github.com/dadrus/heimdall/blob/main/internal/cache/memory/cache_test.go
* What should we do, if calculation of memory size for a given item fails

swithek commented on cache.go:
```diff
@@ -7,6 +7,7 @@ import (
 	"sync"
 	"time"
 
+	"github.com/DmitriyVTitov/size"
```
We should avoid using external libraries.

dadrus replied:
dependency removed

swithek commented on options.go:
```diff
@@ -18,6 +18,7 @@ func (fn optionFunc[K, V]) apply(opts *options[K, V]) {
 // options holds all available cache configuration options.
 type options[K comparable, V any] struct {
 	capacity              uint64
+	sizeInBytes           uint64
```
`totalCost`

dadrus replied:
I have renamed the property as you suggested

swithek commented on options.go:
```diff
@@ -75,3 +76,11 @@ func WithDisableTouchOnHit[K comparable, V any]() Option[K, V] {
+
+// WithMemorySize sets the maximum memory size the cache is allowed to grow.
+// If used together with WithCapacity, WithMemorySize overrules the maximum capacity.
+func WithMemorySize[K comparable, V any](s uint64) Option[K, V] {
```
```suggestion
func WithTotalCost[K comparable, V any](s uint64) Option[K, V] {
```

dadrus replied:
Option has been renamed

swithek reviewed (changes requested):
Hi, thanks for the PR! And sorry for the delayed review.

My main suggestion would be to change the name/concept of this feature from "size" to "weight" or "cost". This way this new functionality could be a bit more versatile: some users might want to track memory usage, others -- some other piece of data that determines how their resources are/should be used.

For this to work properly, we would need another functional option: `WithCostFunc(fn func(item))` (or something similar). The provided function would be used to determine the cost/weight (or size) of every item. This way the users would be able to use external libraries without introducing them into our codebase as well.

swithek commented:
As for the tests: if the improvements are trivial, feel free to implement them, if not, then just ensure that the new functionality is properly covered.

dadrus committed:
options updated to implement the suggested new option

dadrus committed:
cache impl updated + tests

dadrus committed:
dependencies updated

dadrus commented:
@swithek: You're very welcome :) and excuse me for the belated answer and the updates as well.

I’ve updated the implementation based on your suggestion, with a couple of deviations, which are close to the suggestion in https://github.com/jellydator/ttlcache/issues/135#issuecomment-2090449169:  

The `WithTotalCost` function signature and the cost calculation function differ slightly:  
   - The cost calculation function now accepts two parameters—the key and the item—since both influence the total memory usage in my case.  
   - `WithTotalCost` takes the cost calculation function as its second parameter. Adding a separate option for setting only the calculation function would leave the configuration incomplete.  

Additionally, there are further other points worth discussing:  

* With this new functionality, `WithCapacity` could potentially be replaced by the new option. For instance, it could be implemented internally as:  
    ```go
    func WithCapacity[K comparable, V any](c uint64) Option[K, V] {
        return WithTotalCost(c, func(key K, item V) uint64 { return 1 })
    }
    ```  
   Alternatively, it might even be removed entirely to simplify the codebase, though this could be a future consideration.  

* As you can see, I’ve added additional test cases to cover the new functionality. However, I was unable to redesign the entire test function, as it’s unclear what the specific expectations are for each of the existing test cases. If you could provide comments describing the intended expectations for each case, I’d be glad to work on a complete redesign of the test function to improve clarity and maintainability.

* Personally, I find the name `WithMaxCost` or `WithMaxWeight` (as suggested by @davseby  in https://github.com/jellydator/ttlcache/issues/135#issuecomment-2090449169) more concise and descriptive for the new option, though that’s subjective.  

While I don’t think the first two points need to be addressed in this PR, I’d appreciate your feedback on all of them.

dadrus committed:
imports organized

dadrus committed:
readme updated to document the new configuration option

dadrus committed:
better explanation in the example

dadrus committed:
readme line removed by accident restored

dadrus committed:
function renamed (typo fixed)

dadrus committed:
wording fixed

dadrus committed:
useless sentence removed

swithek commented on cache.go:
```diff
@@ -36,6 +37,7 @@ type Cache[K comparable, V any] struct {
 
 		timerCh chan time.Duration
 	}
+	costs uint64
```
```suggestion
	cost uint64
```

dadrus replied:
Renamed

swithek commented on cache_test.go:
```diff
@@ -1228,9 +1258,17 @@ func Test_SuppressedLoader_Load(t *testing.T) {
+		c.options.costsCalcFunc = func(key string, item string) uint64 {
+			// 72 bytes are used by the Item struct
+			// 2 * 16 bytes are used by the used string headers (key and item)
+			return uint64(104 + len(key) + len(item))
```
Would it be possible to use something simpler here? Something unrelated to an item size? For example: just `len(key)`.

dadrus replied:
For tests, it doesn't indeed matter :)

dadrus replied:
simplified

swithek commented on options.go:
```diff
@@ -15,9 +15,15 @@ func (fn optionFunc[K, V]) apply(opts *options[K, V]) {
 type options[K comparable, V any] struct {
 	capacity              uint64
+	totalCost             uint64
+	costsCalcFunc         CostCalcFunc[K, V]
```
```suggestion
	costFunc         CostFunc[K, V]
```

dadrus replied:
renamed

swithek commented on options.go:
```diff
@@ -15,9 +15,15 @@ func (fn optionFunc[K, V]) apply(opts *options[K, V]) {
 
+// CostCalcFunc is used to calculate the cost of the key and the item to be
+// inserted into the cache.
+type CostCalcFunc[K comparable, V any] func(key K, item V) uint64
```
```suggestion
type CostFunc[K comparable, V any] func(item *Item[K, V]) uint64
```

Is there a reason why we can't have the actual `*Item[K, V]` as a parameter here? With it, the API would be more consistent and predictable.

swithek replied:
The only complication that I can see is in the `set()` method: when you are creating the `oldValue` variable, you might need to clone the whole item, otherwise the same updated item/pointer will be used when calculating old and new costs.

If you need to clone an item there, it might be a good idea to add a `Clone()` method to the item type (it should just recreate the item with the same fields while the item's mutex is locked -- it's the same logic as in any other exported Item method).

dadrus replied:
hm... I don't think cloning an item is a good idea - it might be expensive. The idea I'm thinking of instead, is adding a new `cost` property to the `Item` struct and reuse it.

dadrus replied:
The last commit includes a change to the signature of the CostFunc function as you suggested. Instead of holding a cost of each item, I moved the calculation of the old cost to a line before the item is updated for now.

dadrus replied:
Indeed, I’d prefer avoiding the recalculation of old costs during updates and instead storing the calculated cost directly in the item. When a new item is created, it would invoke the `CostFunc` to compute and store the cost, and the same would happen during an update. However, this approach would require modifying the `NewItem` function, which would be a breaking change. Honestly, it also raises the question of why the `NewItem` function is public in the first place?

This way the update part would look like:

```go
if elem != nil {
	// update/overwrite an existing item
	item := elem.Value.(*Item[K, V])
	oldItemCosts := item.cost

	item.update(value, ttl)

	c.updateExpirations(false, elem)

	if c.options.maxCost != 0 {
		c.cost = c.cost - oldItemCosts + item.cost

		for c.cost > c.options.maxCost {
			c.evict(EvictionReasonTotalCostExceeded, c.items.lru.Back())
		}
	}

	return item
}
```

and the code for adding new item would become

```go
// create a new item
var costFunc CostFunc
if c.options.maxCost != 0 {
	costFunc = c.options.costFunc
} else {
	costFunc = func(_ *Item) uint64 { return 0 }
}

item := NewItem(key, value, ttl, c.options.enableVersionTracking, costFunc)
elem = c.items.lru.PushFront(item)
c.items.values[key] = elem
c.updateExpirations(true, elem)

if c.options.maxCost != 0 {
	c.cost += item.cost

	for c.cost > c.options.maxCost {
		c.evict(EvictionReasonTotalCostExceeded, c.items.lru.Back())
	}
}
```

The for loop in the evict function would also be simpler and faster

```go
for i := range elems {
	item := elems[i].Value.(*Item[K, V])
	delete(c.items.values, item.key)

	if c.options.maxCost != 0 {
		c.cost -= item.cost
	}

	c.items.lru.Remove(elems[i])
	c.items.expQueue.remove(elems[i])

	for _, fn := range c.events.eviction.fns {
		fn(reason, item)
	}
}
```

swithek replied:
> hm... I don't think cloning an item is a good idea - it might be expensive.

Yup, that's a good point.

> When a new item is created, it would invoke the CostFunc to compute and store the cost, and the same would happen during an update.

That would probably be more accurate as well, since the value inside the item might change after creation (i.e., the user would still have access to the value and they might change it from the outside).

> However, this approach would require modifying the NewItem function, which would be a breaking change. Honestly, it also raises the question of why the NewItem function is public in the first place?

Yes, that's a good question... For the longest time, this function was private and the reasons for that are similar to what you said/are trying to add. However, in #151 it was made public, which is not a bad change as it serves a purpose and has a use case, but it could have been implemented a bit better: just like the Set/Get methods, it should probably accept functional options as a parameter to update its internal fields. Unfortunately, we cannot change the signature of this function before the next major release, so we have two options:
- Create another `NewItem` function (e.g., `NewItemWithOpts`) and accept functional options, and then deprecate the old `NewItem`. It would also be removed in the next major release.
- Create a private `NewItem` alternative until `NewItemWithOpts` is added. I don't like this approach since it would undermine the whole purpose of #151. However, for now, to finish this feature, it might suffice. `NewItemWithOpts` could be added in a separate PR.

swithek replied:
@dadrus let me know when you'd like me to review your changes.

dadrus replied:
Will do so. I'm pretty swamped this week, but hope to finalize the PR during the WE.

Most of the changes, we were discussing are already implemented. Some of the tests require however updates

swithek commented on cache.go:
```diff
@@ -137,7 +139,20 @@ func (c *Cache[K, V]) set(key K, value V, ttl time.Duration) *Item[K, V] {
+			}
+		}
+
 		c.updateExpirations(false, elem)
```
To keep things consistent with your change below (in the same method), can you move this `updateExpirations` call to before the new if?

dadrus replied:
done

swithek commented on options.go:
```diff
@@ -15,9 +15,15 @@ func (fn optionFunc[K, V]) apply(opts *options[K, V]) {
 // options holds all available cache configuration options.
 type options[K comparable, V any] struct {
 	capacity              uint64
+	totalCost             uint64
```
```suggestion
	maxCost             uint64
```

dadrus replied:
renamed

swithek commented on options.go:
```diff
@@ -75,3 +81,14 @@ func WithDisableTouchOnHit[K comparable, V any]() Option[K, V] {
+// The actual cost calculation for each inserted item happens by making use of the
+// callback CostCalcFunc.
+// If used together with WithCapacity, WithTotalCost overrules the maximum capacity.
+func WithTotalCost[K comparable, V any](s uint64, callback CostCalcFunc[K, V]) Option[K, V] {
```
```suggestion
func WithMaxCost[K comparable, V any](s uint64, callback CostFunc[K, V]) Option[K, V] {
```
As you suggested :+1:

dadrus replied:
renamed

swithek commented on options.go:
```diff
@@ -75,3 +81,14 @@ func WithDisableTouchOnHit[K comparable, V any]() Option[K, V] {
+// WithTotalCost sets the maximum cost the cache is allowed to use (e.g. the used memory).
+// The actual cost calculation for each inserted item happens by making use of the
+// callback CostCalcFunc.
+// If used together with WithCapacity, WithTotalCost overrules the maximum capacity.
```
How? The capacity check and eviction takes place before the cost check and eviction? I don't think we should mention that these two checks clash and override each other when, it seems, they can be safely used together.

dadrus replied:
I see your point. I've removed the problematic sentence from the description. What I aimed to convey is that WithMaxCost allows setting an upper limit on the total cost of managed items: Even when there is remaining capacity for additional items, older items may still be evicted if the maximum cost threshold is surpassed.

swithek reviewed (changes requested):
@dadrus these are some good additions :+1: Just a few more changes and I think it should be ready to be merged.

swithek commented:
> With this new functionality, WithCapacity could potentially be replaced by the new option. For instance, it could be implemented internally as: [...]
> Alternatively, it might even be removed entirely to simplify the codebase, though this could be a future consideration.

If the new `WithMaxCost` is used the same way `WithCapacity` is, then yes, `WithCapacity` becomes redundant. However, what if someone wanted to have a capacity check and an additional cost (e.g., size, as in your use case) check? The overlap between these two features isn't complete, so removing one of them, or even integrating one of them into another isn't a very good option.

> As you can see, I’ve added additional test cases to cover the new functionality. However, I was unable to redesign the entire test function, as it’s unclear what the specific expectations are for each of the existing test cases. If you could provide comments describing the intended expectations for each case, I’d be glad to work on a complete redesign of the test function to improve clarity and maintainability.

That's perfectly fine, the redesign of the test suite isn't a critical task for now (it's not even a task at the moment), so no need spend too much time on it, especially in this PR. As for expectations of each test case, I don't think I can add much at the moment either; each case has a decent description, so that should give a big enough clue.

dadrus committed:
costs renamed to cost

dadrus committed:
cost used in tests simplified

dadrus committed:
CostCalcFunc renamed to CostFunc

dadrus committed:
costsCalcFunc renamed to costFunc

dadrus committed:
updateExpirations moved

dadrus committed:
totalCost renamed to maxCost

dadrus committed:
WithTotalCost option renamed to WithMaxCost

dadrus committed:
example in README updated

dadrus committed:
signature of the WithMaxCost option changed

dadrus committed:
Merge branch 'v3' into feat/cache_size_in_bytes

dadrus committed:
readme updated

dadrus committed:
Item impl updated to hold the current cost and the corresponding cost calculation function; new constructor expecting options

dadrus committed:
cache impl updated to make use of the new item properties

dadrus committed:
new item options

dadrus committed:
tests for the new options

dadrus committed:
new Item related tests

dadrus committed:
cache tests fixed to make them compile

dadrus committed:
item option implementation updated to become a private interface

dadrus committed:
cache set test updated

dadrus commented:
@swithek: Please excuse me for the longer silence. The new updates implement the approach discussed in https://github.com/jellydator/ttlcache/pull/152#discussion_r1842184242 and ff. Please review

swithek commented on options.go:
```diff
@@ -75,3 +80,49 @@ func WithDisableTouchOnHit[K comparable, V any]() Option[K, V] {
+}
+
+// WithCostFunc configures the cost calculation function for an item
+func WithCostFunc[K comparable, V any](costFunc CostFunc[K, V]) ItemOption[K, V] {
```
Make this private for now.

dadrus replied:
done

swithek commented on options.go:
```diff
@@ -75,3 +80,49 @@ func WithDisableTouchOnHit[K comparable, V any]() Option[K, V] {
+// WithVersionTracking deactivates ot activates item version tracking.
+// If version tracking is disabled, the version is always -1.
+// It has no effect when used with Get().
+func WithVersionTracking[K comparable, V any](enable bool) ItemOption[K, V] {
```
Make this private for now.

dadrus replied:
done

swithek commented on options.go:
```diff
@@ -75,3 +80,49 @@ func WithDisableTouchOnHit[K comparable, V any]() Option[K, V] {
+}
+
+// ItemOption represents an option to be applied to an Item on creation
+type ItemOption[K comparable, V any] interface {
```
Make this private for now.

dadrus replied:
done

swithek commented on item.go:
```diff
@@ -30,25 +30,34 @@ type Item[K comparable, V any] struct {
+}
+
+// NewItemWithOpts creates a new cache item.
+func NewItemWithOpts[K comparable, V any](key K, value V, ttl time.Duration, opts ...ItemOption[K, V]) *Item[K, V] {
```
Make this private for now.

dadrus replied:
done

swithek commented on item.go:
```diff
@@ -30,25 +30,34 @@ type Item[K comparable, V any] struct {
 }
 
 // NewItem creates a new cache item.
+// Deprecated, use NewItemWithOpts instead
```
Let's remove this for now.

dadrus replied:
done

swithek commented on item.go:
```diff
@@ -30,25 +30,34 @@ type Item[K comparable, V any] struct {
 	}
 
-	if !enableVersionTracking {
-		item.version = -1
```
The default version value (-1) should still be set, because the version option might not be provided at all.

dadrus replied:
But `-1` is not the default value for the `version`. It is (and also was before this PR) only set to `-1` if version tracking is disabled. This is now covered by the new `withVersionTracking` option, which is always used by the updated implementation of the `NewItem` function.

dadrus replied:
There was a bit of confusion regarding that option. I thought version tracking was enabled by default.

swithek commented on item.go:
```diff
@@ -62,6 +71,7 @@ func (item *Item[K, V]) update(value V, ttl time.Duration) {
 	defer item.mu.Unlock()
 
 	item.value = value
+	item.cost = item.calculateCost(item)
```
The cost needs to be set in `NewItemWithOpts` as well. `update()` isn't used there.

dadrus replied:
By default, if the `WithMaxCost` function is not used, the cost is always `0` and the default implementation of `calculateCost` just returns `0`. You can see that in the `newItemWithOpts` implementation.

If the `WithMaxCost` option is used, it will result in a call to `withCostFunc`, which will set the `costFunc` on the item and also call it. I'll move this call to the implementation of the `newItemWithOpts` function.

dadrus replied:
https://github.com/jellydator/ttlcache/pull/152/commits/5cb43bf6bac41c894f108681075e00dcc1f22484 made it more clear

swithek replied:
Ah yes, I missed that part. I think your updated approach is clearer indeed :+1:

swithek commented on cache.go:
```diff
@@ -15,6 +15,7 @@ const (
 	EvictionReasonDeleted EvictionReason = iota + 1
 	EvictionReasonCapacityReached
 	EvictionReasonExpired
+	EvictionReasonTotalCostExceeded
```
```suggestion
	EvictionReasonMaxCostExceeded
```

To keep things consistent with the options.

dadrus replied:
done

swithek reviewed (commented):
Looks good, left a few comments.

I think for now it's best to make the new item func and its options private (marked by comments). I'll come back to this myself once this PR is merged. There are a few other features in the queue that might benefit from this, so I'd like to take everything into consideration and revise the signatures if needed before finalising these types/funcs. Besides, this PR is large enough already.

dadrus committed:
made new item related functions private + eviction reason renamed

dadrus committed:
moved initial item cost calsulation to the constructor function

dadrus committed:
test functions renamed to reflect the private nature of functions under test

dadrus committed:
version tracking fixed

dadrus commented:
@swithek: All your requests have been addressed. As you may have noticed, I’ve also updated the title and description of this PR. Please have another look.

dadrus committed:
small readme update

dadrus commented on cache.go:
```diff
@@ -153,11 +166,19 @@ func (c *Cache[K, V]) set(key K, value V, ttl time.Duration) *Item[K, V] {
 	c.items.values[key] = elem
 	c.updateExpirations(true, elem)
 
+	if c.options.maxCost != 0 {
```
Same as question as above

swithek replied:
Could you clarify which question you're referring to?

swithek replied:
I'll merge the PR once we resolve this conversation.

swithek replied:
Not sure if your comment got deleted or something, but it seems your question was as follows (I found this in a github notification email):
> Just a comment: most probably the if clause can be removed (just retaining the contents of it). The addition does not have any performance impact and the for loop will not be executed anyway if maxCost is not set, as both the cost and maxCost will always be 0 in such case. What do you think?

I think we should keep these ifs. It makes the intention of this implementation a bit clearer and adds an additional safety check.

swithek reviewed (approved):
I think this is ready to be merged. @dadrus thank you for your patience and all the revisions you made. This is probably the biggest PR in recent months, so congrats on successfully tackling this challenge and finishing it!

I'll squash-merge this to the v3 branch, so it will be possible to download it from there directly. However, the next major version (v3.4) will be released in a few weeks, since there's a plan to include more features in it.

dadrus commented:
I had to delete some of my "review" comments that you responded to in https://github.com/jellydator/ttlcache/pull/152#discussion_r1866634560 because gh wouldn't let me resolve them otherwise.  

I don’t have any additional topics to bring up, but I want to thank you for your patience and the many reviews you’ve provided to help shape this PR into what it is now. 

If you don’t have anything else either, let’s go ahead and squash it :)
